### ISSUE TITLE
Z3v4.5 forgets conclusions of autolemmas triggered by terms constructed while executing an assert statement

### ISSUE THREAD

**Issue by necto** [verifast/verifast#157]
I need to verify the assertion on the last line:
```C
/*@
lemma_auto(length(l) == 2) void list_of_two(void* p);
requires uchars(p, 2, ?l);
ensures uchars(p, 2, l) &*& length(l) == 2 &*& l == cons(_, cons(_, _));
@*/
void to_verify()
/*@ requires true; @*/ 
/*@ ensures true; @*/
{
  unsigned char x[2];
  x[0] = 0;
  //@ assert uchars(x, 2, ?lll);
  //@ assert length(lll) == 2;
  //@ assert lll != nil;
  // @ list_of_two(x);
  //@ assert lll == cons(_, cons(_, _));
}
```

The code is autogenerated, and it is difficult to insert a lemma like `list_of_two` where it is commented out currently. So automatic lemma insertion mechanism, really saved a lot of effort in similar situations. In this case, though, it fails to execute when I use the `-prover z3v4.5`:

>$ verifast -prover z3v4.5 -c tst.c 
tst.c
tst.c(16,29-33): Cannot prove that 'value0' is not an instance of constructor 'nil'
 
Notably, the default solver `redux` seems to have no problem with this example:

>$ verifast -c tst.c 
tst.c
0 errors found (6 statements verified)

Unfortunately, I need z3 for other pieces of the code, where redux fails short.
Why does not verifast trigger the lemma when run with z3 solver, and is there a trick to force it?

**Comment by btj** [comment]
Your autolemma has nothing to do with it; it is not applied in either case. The following version also verifies with Redux and fails with Z3v4.5:
```c
void to_verify()
/*@ requires true; @*/ 
/*@ ensures true; @*/
{
  unsigned char x[2];
  x[0] = 0;
  //@ assert uchars(x, 2, ?lll);
  //@ assert length(lll) == 2;
  //@ assert lll == cons(_, cons(_, _));
}
```
And so does the following version:
```c
/*@

lemma void length_update_corollary<t>(int i, t x, list<t> xs)
    requires length(xs) == 2;
    ensures update(i, x, xs) == cons(_, cons(_, _));
{
    assert length(update(i, x, xs)) == 2;
}

@*/
```
It seems that both Redux and Z3 apply the autolemma update_length defined in list.h, but Redux remembers its conclusion and Z3 forgets it.

The following version, however, verifies with both:
```c
/*@

lemma void note(bool b)
    requires b;
    ensures b;
{}

lemma void length_update_corollary<t>(int i, t x, list<t> xs)
    requires length(xs) == 2;
    ensures update(i, x, xs) == cons(_, cons(_, _));
{
    note(length(update(i, x, xs)) == 2);
}

@*/
```
I could fix this by making `assert` behave like `note`, but I hesitate a bit because it might have a performance overhead. (That's easy to check, though.) Let me know if you'd like me to give that a try. Of course, if you could generate `note` calls instead of `assert` statements, that would fix it to.

**Comment by btj** [comment]
BTW: I have a theory for what's happening. Redux applies axioms eagerly, so in this case it applies `length_update`, as soon as the term `length(update(i, x, xs))` is constructed, before the `assert` condition is proved by contradiction. Z3v4.5 probably applies them lazily, so it applies `length_update` after the assumption that `length(update(i, x, xs)) != 2` is pushed. When that assumption is popped again, the instantiated axiom is popped along with it.

**Comment by necto** [comment]
In my generated code the assertion about the length in not present directly
either, but rather a part of a bigger precise predicate that should auto
close. So it would be non trivial to insert a `note` there.
Would your idea of unifying `assert` and `note` help in this case (when
assertion is not explicitly present)?

If the change is simple, could you give me a patch, so that I would use it
in my branch?
 In my application performance is secondary

Le ven. 22 févr. 2019 à 23:17, Bart Jacobs <notifications@github.com> a
écrit :

**Comment by btj** [comment]
Fixed by daf8aaf3e7d83f669f0cf006975ebbdafe3d7e2f

**Comment by necto** [comment]
That works. Thank you!
